File: src/main/icon.js
'use strict';

const zlib = require('zlib');

/**
 * Générateur d'icônes de barre d'état, en PNG, sans aucune dépendance ni
 * fichier binaire versionné.
 *
 * Pourquoi ne pas simplement livrer un .png ? Parce que l'icône doit refléter
 * l'état : sur Windows et Linux, la barre d'état n'affiche pas de texte à côté
 * de l'icône (contrairement à macOS et `setTitle`). Le seul moyen d'y donner
 * une information d'un coup d'œil est donc de la DESSINER dans l'icône. On
 * peint ici une petite jauge dont le remplissage suit la consommation.
 */

const CRC_TABLE = (() => {
  const t = new Int32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
    t[n] = c;
  }
  return t;
})();

function crc32(buf) {
  let c = -1;
  for (let i = 0; i < buf.length; i++) c = CRC_TABLE[(c ^ buf[i]) & 0xff] ^ (c >>> 8);
  return (c ^ -1) >>> 0;
}

function chunk(type, data) {
  const len = Buffer.alloc(4);
  len.writeUInt32BE(data.length, 0);
  const body = Buffer.concat([Buffer.from(type, 'ascii'), data]);
  const crc = Buffer.alloc(4);
  crc.writeUInt32BE(crc32(body), 0);
  return Buffer.concat([len, body, crc]);
}

/** Encode un buffer RGBA (w*h*4) en PNG. */
function encodePng(rgba, w, h) {
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(w, 0);
  ihdr.writeUInt32BE(h, 4);
  ihdr[8] = 8; // 8 bits par canal
  ihdr[9] = 6; // RGBA
  // Chaque scanline est préfixée de son octet de filtre (0 = aucun).
  const raw = Buffer.alloc((w * 4 + 1) * h);
  for (let y = 0; y < h; y++) {
    raw[y * (w * 4 + 1)] = 0;
    rgba.copy(raw, y * (w * 4 + 1) + 1, y * w * 4, (y + 1) * w * 4);
  }
  return Buffer.concat([
    Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]),
    chunk('IHDR', ihdr),
    chunk('IDAT', zlib.deflateSync(raw, { level: 9 })),
    chunk('IEND', Buffer.alloc(0)),
  ]);
}

/**
 * Dessine le symbole TRACE : un trait d'enregistreur graphique, surmonté
 * d'une jauge de remplissage.
 *
 * @param {number} size      côté en pixels (22 en 1x, 44 en 2x)
 * @param {number} fill      0..1, part de la jauge remplie (null => pas de jauge)
 * @param {boolean} template macOS : icône monochrome noir+alpha, teintée par le système
 * @param {Array} tint       couleur RGB hors mode template
 */
function drawTrayIcon(size, fill = null, template = true, tint = [255, 180, 84]) {
  const rgba = Buffer.alloc(size * size * 4, 0);
  const s = size / 22; // échelle depuis la grille de référence 22x22
  const [r, g, b] = template ? [0, 0, 0] : tint;

  const px = (x, y, a) => {
    x = Math.round(x);
    y = Math.round(y);
    if (x < 0 || y < 0 || x >= size || y >= size) return;
    const i = (y * size + x) * 4;
    const prev = rgba[i + 3];
    const alpha = Math.max(prev, Math.min(255, Math.round(a * 255)));
    rgba[i] = r;
    rgba[i + 1] = g;
    rgba[i + 2] = b;
    rgba[i + 3] = alpha;
  };

  const rect = (x0, y0, w, h, a = 1) => {
    for (let y = 0; y < h; y++) for (let x = 0; x < w; x++) px(x0 + x, y0 + y, a);
  };

  // --- Le tracé : une courbe échantillonnée, comme un stylet sur le papier.
  const points = [0.08, 0.5, 0.26, 0.82, 0.44, 0.14, 0.63, 0.66, 0.92, 0.34];
  const traceTop = 3.5 * s;
  const traceH = 11 * s;
  const thickness = Math.max(2, Math.round(2.4 * s));

  for (let i = 0; i < points.length - 2; i += 2) {
    const x1 = points[i] * size;
    const y1 = traceTop + (1 - points[i + 1]) * traceH;
    const x2 = points[i + 2] * size;
    const y2 = traceTop + (1 - points[i + 3]) * traceH;
    const steps = Math.max(2, Math.ceil(Math.hypot(x2 - x1, y2 - y1) * 3));
    for (let k = 0; k <= steps; k++) {
      const t = k / steps;
      const x = x1 + (x2 - x1) * t;
      const y = y1 + (y2 - y1) * t;
      // Trait épaissi verticalement ET horizontalement : sur une pente raide,
      // un épaississement uniquement vertical laisse la ligne maigre.
      for (let d = 0; d < thickness; d++) {
        px(x, y + d, 1);
        px(x + 0.5, y + d, 1);
      }
    }
  }

  // --- La jauge d'état, en pied d'icône.
  if (fill != null) {
    const barY = Math.round(17.5 * s);
    const barH = Math.max(1, Math.round(1.8 * s));
    const barX = Math.round(1.5 * s);
    const barW = Math.round(19 * s);
    rect(barX, barY, barW, barH, 0.22); // rail
    const filled = Math.round(barW * Math.max(0, Math.min(1, fill)));
    if (filled > 0) rect(barX, barY, filled, barH, 1);
  }

  return encodePng(rgba, size, size);
}

/**
 * Icône applicative : une tuile graphite au coin arrondi, traversée par le
 * tracé ambre. Même vocabulaire que l'icône de barre d'état, mais avec un
 * fond — une icône de Dock ou de menu Démarrer doit tenir sur n'importe quel
 * papier peint.
 */
function drawAppIcon(size = 1024) {
  const rgba = Buffer.alloc(size * size * 4, 0);
  const radius = size * 0.22;
  const bg = [20, 22, 26];
  const trace = [255, 180, 84];

  const set = (x, y, c, a) => {
    const i = (y * size + x) * 4;
    // Composition « source-over » sur ce qui est déjà peint.
    const sa = a;
    const da = rgba[i + 3] / 255;
    const out = sa + da * (1 - sa);
    if (out <= 0) return;
    for (let k = 0; k < 3; k++) rgba[i + k] = Math.round((c[k] * sa + rgba[i + k] * da * (1 - sa)) / out);
    rgba[i + 3] = Math.round(out * 255);
  };

  // Fond : rectangle à coins arrondis, avec un lissage sur le rayon.
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const dx = Math.max(radius - x, x - (size - radius), 0);
      const dy = Math.max(radius - y, y - (size - radius), 0);
      const d = Math.hypot(dx, dy);
      const a = d <= radius - 1 ? 1 : d >= radius + 1 ? 0 : (radius + 1 - d) / 2;
      if (a > 0) set(x, y, bg, a);
    }
  }

  // Lignes de grille du papier.
  for (let i = 1; i < 4; i++) {
    const y = Math.round(size * (0.3 + i * 0.12));
    for (let x = Math.round(size * 0.14); x < size * 0.86; x++) set(x, y, [255, 255, 255], 0.07);
  }

  // Le tracé.
  const pts = [0.14, 0.42, 0.3, 0.66, 0.44, 0.28, 0.58, 0.58, 0.72, 0.36, 0.86, 0.52];
  const w = Math.max(2, size * 0.035);
  for (let i = 0; i < pts.length - 2; i += 2) {
    const x1 = pts[i] * size;
    const y1 = size * (0.78 - pts[i + 1] * 0.5);
    const x2 = pts[i + 2] * size;
    const y2 = size * (0.78 - pts[i + 3] * 0.5);
    const steps = Math.ceil(Math.hypot(x2 - x1, y2 - y1) * 2);
    for (let k = 0; k <= steps; k++) {
      const t = k / steps;
      const cx = x1 + (x2 - x1) * t;
      const cy = y1 + (y2 - y1) * t;
      const r = w / 2;
      for (let oy = -Math.ceil(r); oy <= Math.ceil(r); oy++) {
        for (let ox = -Math.ceil(r); ox <= Math.ceil(r); ox++) {
          const d = Math.hypot(ox, oy);
          if (d > r + 0.7) continue;
          const px2 = Math.round(cx + ox);
          const py2 = Math.round(cy + oy);
          if (px2 < 0 || py2 < 0 || px2 >= size || py2 >= size) continue;
          set(px2, py2, trace, d <= r - 0.5 ? 1 : Math.max(0, r + 0.7 - d) / 1.2);
        }
      }
    }
  }

  return encodePng(rgba, size, size);
}

module.exports = { drawTrayIcon, drawAppIcon, encodePng };

File: src-tauri/src/icon.rs
//! Générateur d'icônes de barre d'état, en PNG, sans aucune dépendance
//! graphique ni fichier binaire versionné.
//!
//! Pourquoi ne pas simplement livrer un `.png` ? Parce que l'icône doit
//! refléter l'état. Sur Windows et Linux, la barre d'état n'affiche pas de
//! texte à côté de l'icône — contrairement à macOS et son titre. Le seul moyen
//! d'y donner une information d'un coup d'œil est donc de la DESSINER dans
//! l'icône : une petite jauge dont le remplissage suit la consommation.

use std::io::Write;

/// Table de CRC-32, celle du format PNG.
fn crc32(bytes: &[u8]) -> u32 {
    static TABLE: std::sync::LazyLock<[u32; 256]> = std::sync::LazyLock::new(|| {
        let mut t = [0u32; 256];
        for (n, slot) in t.iter_mut().enumerate() {
            let mut c = n as u32;
            for _ in 0..8 {
                c = if c & 1 != 0 {
                    0xedb8_8320 ^ (c >> 1)
                } else {
                    c >> 1
                };
            }
            *slot = c;
        }
        t
    });
    let mut c = 0xffff_ffffu32;
    for b in bytes {
        c = TABLE[((c ^ *b as u32) & 0xff) as usize] ^ (c >> 8);
    }
    c ^ 0xffff_ffff
}

fn chunk(kind: &[u8; 4], data: &[u8]) -> Vec<u8> {
    let mut out = Vec::with_capacity(data.len() + 12);
    out.extend_from_slice(&(data.len() as u32).to_be_bytes());
    let mut body = Vec::with_capacity(data.len() + 4);
    body.extend_from_slice(kind);
    body.extend_from_slice(data);
    out.extend_from_slice(&body);
    out.extend_from_slice(&crc32(&body).to_be_bytes());
    out
}

/// Encode un tampon RGBA en PNG.
fn encode_png(rgba: &[u8], size: u32) -> Vec<u8> {
    let mut ihdr = Vec::with_capacity(13);
    ihdr.extend_from_slice(&size.to_be_bytes());
    ihdr.extend_from_slice(&size.to_be_bytes());
    ihdr.push(8); // 8 bits par canal
    ihdr.push(6); // RGBA
    ihdr.extend_from_slice(&[0, 0, 0]); // compression, filtre, entrelacement

    // Chaque ligne est préfixée de son octet de filtre (0 = aucun).
    let w = size as usize;
    let mut raw = Vec::with_capacity((w * 4 + 1) * w);
    for y in 0..w {
        raw.push(0);
        raw.extend_from_slice(&rgba[y * w * 4..(y + 1) * w * 4]);
    }

    let mut encoder = flate2::write::ZlibEncoder::new(Vec::new(), flate2::Compression::best());
    let _ = encoder.write_all(&raw);
    let idat = encoder.finish().unwrap_or_default();

    let mut png = vec![0x89, b'P', b'N', b'G', 0x0d, 0x0a, 0x1a, 0x0a];
    png.extend(chunk(b"IHDR", &ihdr));
    png.extend(chunk(b"IDAT", &idat));
    png.extend(chunk(b"IEND", &[]));
    png
}

/// La grille de référence sur laquelle le symbole est dessiné.
const GRID: f64 = 22.0;

/// Dessine le symbole TRACE : un trait d'enregistreur graphique, surmonté
/// d'une jauge de remplissage.
///
/// `fill` va de 0 à 1 ; `None` n'affiche aucune jauge. En mode gabarit,
/// l'icône est noire et ne porte que son alpha : c'est le système qui la
/// teinte selon le thème de la barre, ce qu'aucune couleur figée ne sait
/// faire.
pub fn draw_tray_icon(size: u32, fill: Option<f64>, template: bool) -> Vec<u8> {
    let n = size as usize;
    let mut rgba = vec![0u8; n * n * 4];
    let s = size as f64 / GRID;
    let (r, g, b) = if template { (0, 0, 0) } else { (255, 180, 84) };

    let mut px = |x: f64, y: f64, a: f64| {
        let (xi, yi) = (x.round(), y.round());
        if xi < 0.0 || yi < 0.0 || xi >= size as f64 || yi >= size as f64 {
            return;
        }
        let i = (yi as usize * n + xi as usize) * 4;
        // On garde l'alpha le plus opaque : les segments du tracé se
        // recouvrent, et les composer les épaissirait aux jointures.
        let alpha = rgba[i + 3].max((a.clamp(0.0, 1.0) * 255.0).round() as u8);
        rgba[i] = r;
        rgba[i + 1] = g;
        rgba[i + 2] = b;
        rgba[i + 3] = alpha;
    };

    // --- Le tracé : une courbe échantillonnée, comme un stylet sur le papier.
    const POINTS: [f64; 10] = [0.08, 0.5, 0.26, 0.82, 0.44, 0.14, 0.63, 0.66, 0.92, 0.34];
    let trace_top = 3.5 * s;
    let trace_h = 11.0 * s;
    let thickness = (2.4 * s).round().max(2.0) as i32;

    for i in (0..POINTS.len() - 2).step_by(2) {
        let x1 = POINTS[i] * size as f64;
        let y1 = trace_top + (1.0 - POINTS[i + 1]) * trace_h;
        let x2 = POINTS[i + 2] * size as f64;
        let y2 = trace_top + (1.0 - POINTS[i + 3]) * trace_h;
        let steps = ((x2 - x1).hypot(y2 - y1) * 3.0).ceil().max(2.0) as i32;
        for k in 0..=steps {
            let t = k as f64 / steps as f64;
            let x = x1 + (x2 - x1) * t;
            let y = y1 + (y2 - y1) * t;
            // Épaissi verticalement ET horizontalement : sur une pente raide,
            // un épaississement seulement vertical laisse la ligne maigre.
            for d in 0..thickness {
                px(x, y + d as f64, 1.0);
                px(x + 0.5, y + d as f64, 1.0);
            }
        }
    }

    // --- La jauge d'état, en pied d'icône.
    if let Some(fill) = fill {
        let bar_y = (17.5 * s).round();
        let bar_h = (1.8 * s).round().max(1.0) as i32;
        let bar_x = (1.5 * s).round();
        let bar_w = (19.0 * s).round() as i32;
        let mut rect = |x0: f64, y0: f64, w: i32, h: i32, a: f64| {
            for y in 0..h {
                for x in 0..w {
                    px(x0 + x as f64, y0 + y as f64, a);
                }
            }
        };
        rect(bar_x, bar_y, bar_w, bar_h, 0.22); // le rail
        let filled = (bar_w as f64 * fill.clamp(0.0, 1.0)).round() as i32;
        if filled > 0 {
            rect(bar_x, bar_y, filled, bar_h, 1.0);
        }
    }

    encode_png(&rgba, size)
}
